### PR TITLE
Extract env var value read into a util

### PR DESCRIPTION
**Pull request by orenbm**
We would like to better UT the logic of reading a value from the env
and defaulting to a default value if it's not present. We also have the
logic that transforms this env var value into an int or a duration (and
maybe more in the future) and it is best that this logic is UTed and is
implemented in its own util. This improves the readability of the
config.go file and of the util functions themselves.

This will also let us use these utils in other projects that consume the
authn-client without implementing them there as well.

### CODE REVIEW

File: pkg/authenticator/config/config.go
Comment (sigalsax) [inline pkg/authenticator/config/config.go:38]: in SP I wrote with `time` was there a particular reason why this was changed?

Comment (orenbm) [reply pkg/authenticator/config/config.go:38]: the util receives a string as an input to match the value that will be given in the env. i think that it makes sense that both the value from the env and the default value are of the same type. after we get the value (from the env or default) we convert it to the required type (int, duration, etc.).

File: pkg/utils/environment_variable.go
Comment (sigalsax) [inline pkg/utils/environment_variable.go:43]: why is this needed when you can just get it via `6*time.seconds`?

Comment (orenbm) [reply pkg/utils/environment_variable.go:43]: but what if the value is from the env and not the default value? in that case it will be a string.

Comment (orenbm) [reply pkg/utils/environment_variable.go:43]: that's why i think it's more simple to have them both as strings. this will be the case for every env var regardless to the expected type. The value is always a string and the logic that converts it to the required type is in the util. it will create a uniformity throughout configurable values in this repo and in our other k8s projects.

File: pkg/utils/environment_variable.go
Comment (sigalsax) [inline pkg/utils/environment_variable.go:20]: when would this not be `nil`? an example would help me understand what we are trying to accomplish here (:

Comment (sigalsax) [reply pkg/utils/environment_variable.go:20]: IMO this breaking up of parameters style is really hard for my 👀  to follow

Comment (orenbm) [reply pkg/utils/environment_variable.go:20]: look at the UT :)

Comment (orenbm) [reply pkg/utils/environment_variable.go:20]: this is the way we can pass mocks to the functions in go

File: pkg/utils/environment_variable.go
Comment (sigalsax) [inline pkg/utils/environment_variable.go:27]: return -1 = error or is it return 1?

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:20]: > IMO this breaking up of parameters style is really hard for my 👀 to follow

i understand. i had a conversation about this with @sgnn7 yesterday. the alternative is a line that is too long which is an issue as well.

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:27]: `-1` is the way for me to pass `nil` for an int. it's irrelevant as if `err` is not nil we don't use the returned value.

File: pkg/authenticator/config/config.go
Comment (sigalsax) [inline pkg/authenticator/config/config.go:42]: are there minimum or maximum values for this limit? Can they put `-5` for example? Would the logic hold up against negative numbers?

Comment (sigalsax) [reply pkg/authenticator/config/config.go:42]: In SP we added a check for `envValueInt < minValue` with a minValue being 0 in order to avoid this

File: pkg/utils/environment_variable.go
Comment (sigalsax) [inline pkg/utils/environment_variable.go:1]: In terms of the naming of this package, it doesn't tell me much. So IMO add a comment at the top of the file what benefit this util gives us or change the name. I vote for comment (:

File: pkg/authenticator/config/config.go
Comment (orenbm) [reply pkg/authenticator/config/config.go:42]: hmmmm. do you think it is needed or is it too low level?

Comment (orenbm) [reply pkg/authenticator/config/config.go:42]: do we want to not let users set (for example) `CONJUR_TOKEN_TIMEOUT=-5`?

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:1]: yeah the package name is for all the utils in this project. i will add a comment

Comment (orenbm) [reply pkg/utils/environment_variable.go:1]: Thinking about it again, i think that the file `utils/environment_variable.go` is quite self-explanatory. would the comment `This file centralizes functions for extracting values from environment variables` help the reader understand it better? I think it does not.

Comment (orenbm) [reply pkg/utils/environment_variable.go:1]: i added comments for each method instead

File: pkg/utils/environment_variable.go
Comment (sigalsax) [reply pkg/utils/environment_variable.go:49]: the 6 would be from the env and *time*seconds would be from the code

Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: yuck (sorry)

Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: also - the env should be `6m0s` to match 6 minutes. just `6` won't work.

Comment (sigalsax) [reply pkg/utils/environment_variable.go:49]: But no problem. You gave a good explanation

File: pkg/authenticator/config/config.go
Comment (sigalsax) [reply pkg/authenticator/config/config.go:42]: I can totally see this being an issue _especially_ because it is configurable no? It is coming from an environment and I don't trust user input 🙃

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: would you prefer if the env var would be an int? e.g `CONJUR_TOKEN_TIMEOUT=6`? if so we would change the logic to `strconv.Atoi(envVarValue) * time.Minutes`

File: pkg/authenticator/config/config.go
Comment (orenbm) [reply pkg/authenticator/config/config.go:42]: yes, but it adds another param that the function consumer needs to set where for most of the times it will be `0` and not interesting. i'll let @izgeri give her take on this before going forward.

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: the thing is that this will be a breaking change as today passing an int will not work. However, i believe that this variable was never used correctly and that setting `CONJUR_TOKEN_TIMEOUT=6` does not work also today. verifying now.

Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: yep, this won't work in the current product. i get a `ERROR: 2020/09/10 11:56:22 config.go:128: CAKC010E Failed to parse CONJUR_TOKEN_TIMEOUT. Reason: time: missing unit in duration 6` error.

@izgeri when this env var was created how did you intend to set it? with `6m0s`? do you know who defined this?
do you see a problem with changing this to an int input so it will be `CONJUR_TOKEN_TIMEOUT="6"` instead of `CONJUR_TOKEN_TIMEOUT="6m0s"`?

Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: once we get to a decision i will update the README with the valid value pattern.

File: pkg/authenticator/config/config.go
Comment (izgeri) [inline pkg/authenticator/config/config.go:162]: if we are updating this, can we change the name of this var? it's _not_ the default config, it's the config that the client will use based on the default config _and_ the user-provided config

Comment (orenbm) [reply pkg/authenticator/config/config.go:162]: sure, nice catch.

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: aha, i can see that this env var was used before as:
```
- name: CONJUR_TOKEN_TIMEOUT
+            value: "15s"
``` 
(so it in Slack) so i wouldn't change the UX to be of int to not break changes.

Comment (orenbm) [reply pkg/utils/environment_variable.go:49]: @izgeri can you please review the commit i added with the readme change?

File: pkg/utils/environment_variable.go
Comment (izgeri) [inline pkg/utils/environment_variable.go:11]: what is happening on this line?

Comment (orenbm) [reply pkg/utils/environment_variable.go]: we set the function as a variable so it's clearer to the reader that the function being used in the one defined at the env of the file

File: CHANGELOG.md
Comment (izgeri) [inline CHANGELOG.md:12]:
```suggestion
- Username formatting now correctly only prints `FullUsername` field.
   [cyberark/conjur-authn-k8s-client#126](https://github.com/cyberark/conjur-authn-k8s-client/issues/126))
- Timeout for token retrievals is now correctly reset in consecutive authentication failures.
  [cyberark/conjur-authn-k8s-client#158](https://github.com/cyberark/conjur-authn-k8s-client/issues/158)
```

I am not sure the user impact from these changes is clear

for the first, is it that the log messages now print a more succinct name or a more detailed name? I'd call that out specifically
> Logs now correctly print only the Conjur identity without the policy branch prefix.

for the second option, it's more like
> When authentication fails, the exponential backoff retry is correctly reset so that it will continue to attempt to authenticate until backoff is exhausted.

File: CHANGELOG.md
Comment (izgeri) [inline CHANGELOG.md:16]: What is the user impact of this change? How does it affect the UX?

File: README.md
Comment (izgeri) [inline README.md:41]: this should be broken into multiple lines

> a signed sequence of decimal numbers with optional fraction and unit suffix
I don't know what this means without the example. is there a standard we could link to here?

File: design/templates/solution_design.md
Comment (izgeri) [inline design/templates/solution_design.md:1]: does this indicate your branch may need a rebase?

File: pkg/authenticator/config/config.go
Comment (izgeri) [reply pkg/authenticator/config/config.go:42]: I think we should have error checking to ensure sane and acceptable values are input for this param. is the param currently documented? in the README or just in the CONTRIB? if it's just for dev purposes / dev troubleshooting, we could just leave undocumented for now. if we're adding it AND documenting it, then we need to be more careful (and this would need a changelog entry)

File: CHANGELOG.md
Comment (orenbm) [reply CHANGELOG.md:16]: the client will fail less frequently as we now wait for up to 10 seconds to read the cert instead of immediately reading it. We are also logging our efforts so that's another customer-facing change. would you change the entry according to this?

File: pkg/utils/environment_variable.go
Comment (izgeri) [inline pkg/utils/environment_variable.go:11]: what is happening on this line?

File: pkg/utils/environment_variable.go
Comment (izgeri) [reply pkg/utils/environment_variable.go:30]: might be worth a comment - it's not a pattern I've seen before. maybe other devs would disagree but @sigalsax already had a question :)

File: pkg/utils/environment_variable.go
Comment (izgeri) [reply pkg/utils/environment_variable.go:49]: I'm pretty sure that `CONJUR_TOKEN_TIMEOUT` was added for dev troubleshooting, though I see it is documented in the README

I think it is wise to set a standard and enforce it, we should just make sure that the standard (and the default, btw) is clear

**Review by izgeri — CHANGES_REQUESTED** [review]
left a handful of comments, still would be good for @sgnn7 to weigh in since he chatted with you about some of the implementation details

File: README.md
Comment (orenbm) [reply README.md:41]: i took it from the golang docs for `time`: https://golang.org/pkg/time/#ParseDuration. i'll add a link

Comment (orenbm) [reply README.md:41]: how about `The value should be in a format that can be parsed with [time.ParseDuration](https://golang.org/pkg/time/#ParseDuration) (e.g "6m0s")`?

File: design/templates/solution_design.md
Comment (orenbm) [reply design/templates/solution_design.md:1]: correct

**Comment by orenbm** [comment]
> I think we should have error checking to ensure sane and acceptable values are input for this param. is the param currently documented? in the README or just in the CONTRIB? if it's just for dev purposes / dev troubleshooting, we could just leave undocumented for now. if we're adding it AND documenting it, then we need to be more careful (and this would need a changelog entry)

`CONJUR_CLIENT_CERT_RETRY_COUNT_LIMIT` is for dev purposes so i'm leaving it undocumented. It is verified to be an int value when parsing it.

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:11]: added a comment
```
// The function 'envVarValueOrDefault' is defined below. We set this var at the
// top of the file so when we use the function it is clear that it's the default one.
```

**Comment by orenbm** [comment]
> might be worth a comment - it's not a pattern I've seen before. maybe other devs would disagree but @sigalsax already had a question :)

@izgeri added a comment

**Comment by orenbm** [comment]
> I'm pretty sure that CONJUR_TOKEN_TIMEOUT was added for dev troubleshooting, though I see it is documented in the README

I think it is wise to set a standard and enforce it, we should just make sure that the standard (and the default, btw) is clear

we enforce it in `DurationFromEnvOrDefault`. an error will be raised if the value is not in the valid format.

File: pkg/utils/environment_variable_test.go
Comment (diverdane) [inline pkg/utils/environment_variable_test.go:26]: I think we're deprecating the use of `Convey`? But personally I'm fine with it... it tells the story of what you're testing.

File: pkg/authenticator/config/config.go
Comment (diverdane) [reply pkg/authenticator/config/config.go:42]: I don't want to hold up this PR, but we should really consider doing a followup to allow for minimum/maximum range validation. In most cases, I'd bet that we can define hard-coded values for min and max values that are sensible and are liberal enough to cover all cases. For example, for certificate retries, we could set the minimum to 0 and the maximum to something astronomical (`2**32 - 1`?) If we don't add range checks, I don't think we know what would happen if someone sets retries or some other configured numbered value to -1 (I'm sure it's not in our integration testing).

**Review by diverdane — COMMENTED** [review]
This is good stuff, thanks for moving this logic to a re-usable util package!
I've included some suggestions on how the code might be refactored to simplify and maybe help readability and code coverage.

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:34]: nice, great to learn this!

@sigalsax, @izgeri  ^

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:15]: made it local

File: pkg/utils/environment_variable.go
Comment (sigalsax) [reply pkg/utils/environment_variable.go:34]: So I am understanding correctly if we return an error (for example `return -1, log.RecordedError(log.CAKC010E, envVarName, err.Error())`), no matter what, Golang will zero the return values? So for example, -1 will be 0 because it is an integer?

File: pkg/utils/environment_variable.go
Comment (sigalsax) [reply pkg/utils/environment_variable.go:55]: which is a bit confusing because usually return 0 = succeeding in bash 🙃

File: pkg/utils/environment_variable_test.go
Comment (sigalsax) [reply pkg/utils/environment_variable_test.go:26]: @diverdane what are we moving to if not `Convey`

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:15]: love it, thanks

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:34]: i haven't verified it but in any case we will return 0 to maintain the standard

File: pkg/utils/environment_variable.go
Comment (orenbm) [reply pkg/utils/environment_variable.go:23]: i don't see a use-case where an environment variable holds sensitive data so leaving this out. if we will need it in the future we can add another function that will be called from this one for backwards-compatibility. i don't want to over-complicate this util unless i know that it is needed.

File: pkg/utils/environment_variable_test.go
Comment (orenbm) [reply pkg/utils/environment_variable_test.go:21]: thanks for the thorough example!

File: pkg/authenticator/config/config.go
Comment (orenbm) [reply pkg/authenticator/config/config.go:42]: by adding a minimum value we raise a few questions:
1. what is the minimum value of `CONJUR_TOKEN_TIMEOUT`?
2. what is the minimum value of `CONJUR_CLIENT_CERT_RETRY_COUNT_LIMIT`?
3. what do we do in case the given value is less than the minimum? Do we fail the app or use the default value?
4. does the answer to question #3 changes how we handle cases where the user gave an invalid value for `CONJUR_TOKEN_TIMEOUT`?

while these are good questions, i don't want to ask them in this PR and in this release that we are preparing. Once this is merged and the release is out there, we can look back at these questions and decide properly.

Comment (orenbm) [reply pkg/authenticator/config/config.go:42]: created an issue for this: https://github.com/cyberark/conjur-authn-k8s-client/issues/162

**Review by sigalsax — APPROVED** [review]
LGTM great work!

**Pull request merged**
